Generate box proposals op

samikama opened the pull request:
This PR adds GenerateBoxProposals op for accelerating FPN based networks such as MaskRCNN. It requires #28745 to be merged.

samikama committed:
Adding GenerateBoxProposals op

samikama committed:
Update api_def files and add api_def for the op

samikama commented:
Pinging @tfboyd @tatianashp. This is the third part of ops to accelerate FPN based networks.

aaroey commented:
Hi @alextp and @pengchongjin, would you help to review the API part of this PR? Thanks

alextp commented on tensorflow/core/api_def/base_api/api_def_GenerateBoundingBoxProposals.pbtxt:
```diff
@@ -0,0 +1,17 @@
+      `nms_threshold` intersection-over-union (iou) value, discarding boxes where shorter
+      side is less than `min_size`.
+      Inputs:
+      `scores`: A 4D tensor of shape [Batch, Height, Width, Num Anchors] containing the scores per anchor at given postion
```
Use the in_arg and out_arg fields to populate inputs and outputs: https://github.com/tensorflow/tensorflow/blob/273981699d92c1aa08ca52af99ee187ce5aaa5e1/tensorflow/core/framework/api_def.proto#L79

samikama replied:
This file is auto-generated by tensorflow tools. Do you want me to change them? Wouldn't it break next time they are autogenerated again?

samikama replied:
Done

alextp commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -925,4 +925,38 @@ REGISTER_OP("CombinedNonMaxSuppression")
+    .Input("anchors: float")
+    .Output("rois: float")
+    .Output("roi_probabilities: float")
+    .Attr("pre_nms_topn: int = 6000")
```
Why are these attributes and not inputs?

samikama replied:
Because they don't depend on data? They are model parameters and fixed at the model construction time.

samikama replied:
I converted these to tensors as requested. But considering the op, it will just cause an extra overhead since it will require many constant ops to be scheduled unnecessarily.

alextp commented on tensorflow/tools/api/golden/v1/tensorflow.pbtxt:
```diff
@@ -1296,6 +1296,10 @@ tf_module {
     argspec: "args=[\'params\', \'indices\', \'name\', \'batch_dims\'], varargs=None, keywords=None, defaults=[\'None\', \'0\'], "
   }
+  member_method {
+    name: "generate_bounding_box_proposals"
```
Probably belongs in tf.image, not tf

samikama replied:
Implementation is in tf.image. This is another auto-generated file. What kind of changes should I do so that api generator puts it tf.image?

samikama replied:
It should be done now.

aaroey commented:
@samikama I think the interface of NonMaxSuppressionV2 is fixed, then why does this PR depend on #28745?

aaroey commented:
@pengchongjin, would you help to provide review for the interface from the perspective of object detection?

Thanks.

aaroey commented:
@smit-hinsu, would you please help to look at the implementation after API review is done? Thanks.

aaroey commented:
@wujingyue just in case you're interested in this op.

pengchongjin commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -925,4 +925,54 @@ REGISTER_OP("CombinedNonMaxSuppression")
+      nms_threshold intersection-over-union (iou) value, discarding boxes where shorter 
+      side is less than min_size.
+
+      scores: A 4D tensor of shape [Batch, Height, Width, Num Anchors] containing the scores per anchor at given postion
```
Nit: '.' in the end of the sentence.

samikama replied:
Docs are moved to api_def files.

pengchongjin commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -925,4 +925,54 @@ REGISTER_OP("CombinedNonMaxSuppression")
+      side is less than min_size.
+
+      scores: A 4D tensor of shape [Batch, Height, Width, Num Anchors] containing the scores per anchor at given postion
+      bbox_deltas: is a tensor of shape [Batch, Height, Width, 4 x Num Anchors] boxes encoded to each anchor
```
Nit: '.' in the end of the sentence.

pengchongjin commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -925,4 +925,54 @@ REGISTER_OP("CombinedNonMaxSuppression")
+
+      scores: A 4D tensor of shape [Batch, Height, Width, Num Anchors] containing the scores per anchor at given postion
+      bbox_deltas: is a tensor of shape [Batch, Height, Width, 4 x Num Anchors] boxes encoded to each anchor
+      anchors: A 1D tensor of shape [4 x Num Anchors], representing the anchors.
```
Any particular reason to make Num_Achors be the second dimension?

samikama replied:
It is not 2d. There are 4 times num_anchors elements in the array. Each 4 points representing an anchor box

pengchongjin commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -925,4 +925,54 @@ REGISTER_OP("CombinedNonMaxSuppression")
+      anchors: A 1D tensor of shape [4 x Num Anchors], representing the anchors.
+
+      rois: output RoIs, a 3D tensor of shape [Batch, post_nms_topn, 4], padded by 0 if less than post_nms_topn candidates found.
+      roi_probabilities: probability scores of each roi in 'rois', a 2D tensor of shape [Batch,post_nms_topn], padded with 0 if needed.
```
Are these sorted? Can you comment on it?

pengchongjin commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -925,4 +925,54 @@ REGISTER_OP("CombinedNonMaxSuppression")
+    .Attr("nms_threshold: float = 0.7")
+    .Attr("min_size: float = 16")
+    .Attr("debug: bool = false")
+    .Attr("correct_transform_coords: bool = true")
```
Any comment on what does this do?

aaroey commented on tensorflow/core/kernels/generate_box_proposals_op.cu.cc:
```diff
@@ -0,0 +1,631 @@
+    int num_images, int conv_layer_nboxes, size_t cub_sort_temp_storage_bytes,
+    size_t cub_select_temp_storage_bytes, int nboxes_to_generate, int box_dim) {
+  auto d = context->eigen_gpu_device();
+  TF_RETURN_IF_ERROR(context->allocate_temp(
```
The pattern:
```
context->allocate_temp(...)
SetZero<<<...>>>(...)
```
occurs >10 times, can we add a helper method for that?

samikama replied:
Done

samikama commented:
Sorry for not responding. Vacation and a few other things got in the way. I will get back to this as soon as possible.

samikama committed:
Updating to master

samikama committed:
Fixes for review

samikama committed:
Fix pylint problem

aaroey commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -3888,3 +3888,24 @@ def draw_bounding_boxes(images, boxes, name=None, colors=None):
+
+
+@tf_export("image.generate_bounding_box_proposals")
+def generate_bounding_box_proposals(scores,
```
@alextp could you help to review this change?

alextp replied:
I'm a little swamped; can you find someone else? Or I can look at it on friday.

samikama commented on tensorflow/core/api_def/base_api/api_def_GenerateBoundingBoxProposals.pbtxt:
```diff
@@ -0,0 +1,86 @@
+END
+  }
+  attr {
+    name: "post_nms_topn"
```
This affects the output shape. I was told before that the attributes that change the output shape should be attributes not inputs. If I do it, this will be a data dependent op. Do you really want that?

samikama commented on tensorflow/core/kernels/generate_box_proposals_op.cu.cc:
```diff
@@ -0,0 +1,605 @@
+      d_out_boxes[out_index] = {x1, y1, x2, y2};
+      // d_inout_scores size: (num_images,KA)
+      if (!keep_box)
+        d_inout_scores[image_index * KA + ibox] = FLT_MIN;  // for NMS
```
it is in the system headers. float.h

samikama commented on tensorflow/core/kernels/generate_box_proposals_op.cu.cc:
```diff
@@ -0,0 +1,605 @@
+
+}  // namespace
+
+class GenerateBoundingBoxProposals : public tensorflow::AsyncOpKernel {
```
It was to prevent stream synchronization needed with nms calls, however current event processing infrastructure of TF is insufficient for doing such kind of asynchronous processing thus idea was abandoned. I reverted it back to sync op.

samikama committed:
Address review comments

aaroey commented on tensorflow/core/kernels/generate_box_proposals_op.cu.cc:
```diff
@@ -38,56 +38,66 @@ limitations under the License.
 namespace {
+
+template <bool T>
+__device__ float legacy_op(float);
```
Use camel case for function names, also use more descriptive name e.g. ApplyOffset.

samikama replied:
Done

aaroey commented on tensorflow/core/kernels/generate_box_proposals_op.cu.cc:
```diff
@@ -162,9 +161,12 @@ __global__ void GeneratePreNMSUprightBoxesKernel(
-      // d_inout_scores size: (num_images,KA)
-      if (!keep_box)
-        d_inout_scores[image_index * KA + ibox] = FLT_MIN;  // for NMS
+      // d_scores size: (num_images,KA)
```
Fix the reference to `KA`.

samikama replied:
Done

samikama replied:
Done

aaroey commented on tensorflow/core/kernels/generate_box_proposals_op.cu.cc:
```diff
@@ -99,58 +109,47 @@ __global__ void GeneratePreNMSUprightBoxesKernel(
-      y1 = fmax(fmin(y1, img_height - 1.0f), 0.0f);
-      x2 = fmax(fmin(x2, img_width - 1.0f), 0.0f);
-      y2 = fmax(fmin(y2, img_height - 1.0f), 0.0f);
+      x1 = fmax(fmin(x1, -legacy_op<T>(-img_width)), 0.0f);
```
Applying minus sign twice makes it harder to read, maybe add an `SubtractOffset` func for it.

samikama replied:
Done

samikama replied:
Done

aaroey commented on tensorflow/core/kernels/generate_box_proposals_op.cu.cc:
```diff
@@ -38,56 +38,66 @@ limitations under the License.
-    const float bbox_xform_clip, const bool correct_transform,
+    const float4* d_bbox_deltas, const float4* d_anchors, const int height,
+    const int width, const int num_anchors, const float min_size,
+    const float* d_img_info_vec, const float bbox_xform_clip,
```
Please comment the shape of d_img_info_vec and d_scores.

samikama replied:
They are described in op documentation but I added it here as well.

aaroey commented on tensorflow/core/kernels/generate_box_proposals_op.cu.cc:
```diff
@@ -0,0 +1,594 @@
+          width, num_anchors, min_size, image_info.flat<float>().data(),
+          bbox_xform_clip_default_,
+          reinterpret_cast<float4*>(dev_boxes.flat<float>().data()),
+          nboxes_to_generate, dev_sorted_scores.flat<float>().data(),
```
We already have `dev_boxes_keep_flags` to tell which boxes to keep, why do we need to pass `dev_sorted_scores` and set scores of all not-selected boxes to min(float)?

Also from the code below we'll use `cub::DeviceSelect::Flagged` and `dev_boxes_keep_flags` to filter out those scores, so setting the score to min(float) becomes not useful?

samikama replied:
Hi @aaroey , that is correct. I removed score reference.

samikama replied:
Done

aaroey reviewed (changes requested):
Thanks for the fix, just a few more minor comments.

samikama committed:
Renamed legacy op and added subtract version

samikama commented on tensorflow/core/kernels/generate_box_proposals_op.cu.cc:
```diff
@@ -38,56 +38,66 @@ limitations under the License.
-      // scores from the conv layer, of shape (H,W,A) the num_images dimension
-      // was already removed box_conv_index = a*K + h*W + w
-      const int box_conv_index = d_sorted_scores_keys[image_index * KA + ibox];
+      // scores from the conv layer, of shape (height,width,A) the num_images
```
Done

ppwwyyxx commented:
`correct_transform_coords` is not necessarily correct. The transform logic in this inference op is correct only when it is the inverse of the transform logic during training. 
The history of the "correct_transform_coords", I believe, comes from a [bug in the py-faster-rcnn project](https://github.com/rbgirshick/py-faster-rcnn/blob/781a917b378dbfdedb45b6a56189a31982da1b43/lib/fast_rcnn/bbox_transform.py#L30-L61) (and also many derived projects from it), where the box encoding and decoding are not the exact inverse of each other.
In other words, `correct_transform_coords=True` is correct only when the encoding during training uses the same implementation in [py-faster-rcnn](https://github.com/rbgirshick/py-faster-rcnn/blob/781a917b378dbfdedb45b6a56189a31982da1b43/lib/fast_rcnn/bbox_transform.py#L10-L24), but different implementations do exist. For example, in Facebook we're [dropping](https://github.com/pytorch/pytorch/blob/master/caffe2/operators/generate_proposals_op_util_boxes.h#L43-L81) the old implementation in py-faster-rcnn/Detectron with the `legacy_plus_one` option.

Also I don't know why `correct_transform_coords=False` could be useful in any situation except to match results of legacy buggy implementation. I couldn't understand why an op added to TensorFlow should take care of certain buggy 3rd-party implementation of papers.

aaroey commented:
@samikama I feel the same way, if `correct_transform_coords` is added only for a legacy but buggy usage I don't think we should add that.

mellanox-github commented:
Can one of the admins verify this patch?

pengchongjin reviewed (changes requested):
Interfaces look good to me. Could you please comment on whether input/output scores are sorted?

samikama commented:
@ppwwyyxx and @aaroey Unfortunately, probably due to rush to publication and recognition, authors are publishing buggy code and that code is becoming a reference in the community which sometimes results in new code to be compared against. I believe that buggy code @ppwwyyxx referencing is still quite prevalent in the community and still used same way. I added that flag to be able to satisfy requirements of users still keeping that offset. I will shortly remove it.
@ppwwyyxx is there an errata or public page that shows that offset was a bug that never should have been there in the first place so that if users complain about op behaving differently then we can point them to and ask fixing their models?

ppwwyyxx commented:
The original repo (py-faster-rcnn) is unmaintained. There is only one report I found about this bug  https://github.com/rbgirshick/py-faster-rcnn/issues/852.

It was acknowledged as a bug in https://github.com/facebookresearch/Detectron/issues/298#issuecomment-373942633 (3).

I'm not aware of other resources that mention this problem. But we plan to explain and fix these historical issues together with the next version of Detectron to be released later this year.

samikama commented:
@ppwwyyxx So this bug is not widely known and not really easy to find out. That is in line with my understanding of the situation. @aaroey @alextp are you sure you want to remove compatibility flag, considering that most networks inspired by detectron will likely to have the bug still and will not be able to use this op without changing their networks?

samikama committed:
Removed legacy support

samikama commented:
@tatianashp as discussed in meeting last week, I removed the support for the legacy implementation. If it turned out to be a problem, last commit can be reverted and other changes can be reapplied.
@pengchongjin I added relevant comments to the api_def files since I have been asked to move the documentation to that files above. Inputs scores can be unsorted, op will sort them internally, output will be sorted by scores in descending order.
To be merged after #30893

samikama committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into GenerateBoxProposalsOp

samikama committed:
Change GetCudaStream to GetGpuStream

samikama commented:
@pengchongjin @alextp @aaroey As requested, I removed the legacy mode option. I still believe it is a bad choice since the **bug** is not documented, not fixed, not known by anybody other than a handful of people and still being used in models. I believe removing it will lead to confusion and head scratching. 
Do you have any other concerns?

samikama committed:
Fix api files

samikama commented:
@aaroey, Just pushed the fixes for api files. Thanks

samikama commented:
@aaroey @rthadur please merge this after #30893 since it contains fixes that are used by this op as well.

rthadur commented on tensorflow/tools/api/golden/v1/tensorflow.pbtxt:
```diff
@@ -1340,6 +1340,10 @@ tf_module {
     argspec: "args=[\'params\', \'indices\', \'name\', \'batch_dims\'], varargs=None, keywords=None, defaults=[\'None\', \'0\'], "
   }
+  member_method {
+    name: "generate_bounding_box_proposals"
```
Internal review comments : 'The op needs a visibility:HIDDEN to avoid polluting the toplevel TF namespace'

aaroey replied:
Please fix this @samikama

samikama replied:
Done

pkulzc reviewed (dismissed):
LGTM.

rthadur commented:
@pkulzc  can we pull this in ?

samikama commented:
@rthadur please wait until I make another commit since during the review process, there were some api changes in consumers and at current state it would require unnecessary operations to wire it up.  I will reformat the data layouts to match users.

tensorflowbutler commented:
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

samikama commented:
Hi, Yes this is still being worked on. Waiting for some internal updates.

samikama committed:
Fix review comments and remove obsolete flag that is removed by earlier review modifications

samikama commented:
@aaroey  @pkulzc This should be good to go now.

samikama committed:
Fix buildifier issue

aaroey commented on tensorflow/core/kernels/generate_box_proposals_op.cu.cc:
```diff
@@ -0,0 +1,567 @@
+      tensorflow::OpKernelConstruction* context)
+      : OpKernel(context) {
+    OP_REQUIRES_OK(context, context->GetAttr("post_nms_topn", &post_nms_topn_));
+    CHECK_GT(post_nms_topn_, 0);
```
@samikama remove all CHECKs and return proper error instead. It's not allowed and breaking the merge tests.

samikama replied:
@aaroey Is this a new requirement? I am seeing still there are a lot of references to CHECKs in TF code base. Can you be a bit more specific on which tests it is breaking?

aaroey replied:
This is not new. It's required that new code (since >1 years ago I believed) should avoid checks. The merge process is blocked because of this check.

samikama replied:
That is strange since I can see commits as recent as last month that has same CHECK_* ops in the trunk. Are you sure that it is blocking the merge? If so how are some of the links below passed these tests? Pretty much all of them created after this pr is opened. Perhaps there should be some PR requirements document that everybody should adhere since if we use style/code similar to what is in TF or recent PRs, they are being marked as blockers. Could either somebody fix these in the TF so that other contributors can't use them as reference or can there be some requirements document created that everybody follows that document and prevent confusion and ease up non-google contributors life and improve their enthusiasm.

Thanks,
 
https://github.com/tensorflow/tensorflow/blob/6c32a22396b7ac9f9b9d77b147113281b689b345/tensorflow/stream_executor/rocm/rocm_gpu_executor.cc#L421
https://github.com/tensorflow/tensorflow/blob/6c32a22396b7ac9f9b9d77b147113281b689b345/tensorflow/stream_executor/host/host_gpu_executor.cc#L45
https://github.com/tensorflow/tensorflow/blob/6c32a22396b7ac9f9b9d77b147113281b689b345/tensorflow/stream_executor/gpu/redzone_allocator.cc#L275
https://github.com/tensorflow/tensorflow/blob/6c32a22396b7ac9f9b9d77b147113281b689b345/tensorflow/python/lib/core/py_func.cc#L193
https://github.com/tensorflow/tensorflow/blob/6c32a22396b7ac9f9b9d77b147113281b689b345/tensorflow/python/eager/pywrap_tensor.cc#L121

aaroey reviewed (changes requested):
@samikama remove all CHECKs and return proper error instead. It's not allowed and breaking the merge tests.

alextp commented:
Can this op live in tf-addons?

samikama committed:
Replaced CHECK_OK calls with context returns even though most of the failures were already FATAL.

samikama commented:
Hi, @alextp I don't intend to move this to tf-addons, especially not after more than 6 months of going back on forth in this PR.

samikama committed:
Mark GenerateBoxProposalsOp non-differentiable

rthadur commented on tensorflow/core/kernels/generate_box_proposals_op.cu.cc:
```diff
@@ -0,0 +1,568 @@
+    Tensor dev_boxes_keep_flags;  // bitmask for keeping the boxes or rejecting
+                                  // from output
+    const int nboxes_to_generate = std::min(conv_layer_nboxes, pre_nms_topn);
+    size_t cub_temp_storage_bytes=max(cub_sort_temp_storage_bytes,cub_select_temp_storage_bytes);
```
`third_party/tensorflow/core/kernels/generate_box_proposals_op.cu.cc:398:9: error: no matching function for call to 'max'
        max(cub_sort_temp_storage_bytes, cub_select_temp_storage_bytes);`

can you please check above error

samikama committed:
max->std::max

aaroey committed:
Fix error 'unused variable 'max_postnms_nboxes''

aaroey committed:
Fix error 'unused variable 'max_postnms_nboxes''

aaroey committed:
Add some changes to trigger copybara import again.

ppwwyyxx commented on tensorflow/core/api_def/base_api/api_def_GenerateBoundingBoxProposals.pbtxt:
```diff
@@ -0,0 +1,81 @@
+  in_arg {
+    name: "image_info"
+    description: <<END
+A 2-D float tensor of shape `[num_images, 5]` containing image information Height, Width, Scale.
```
why there are 5 elements each row if it only has height, width and scale? This makes no sense

wbudd replied:
Yeah, [generate_box_proposals_op.cu.cc#L115~118](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/image/generate_box_proposals_op.cu.cc#L115) contains the following...
```C
const float img_height = d_img_info_vec[5 * image_index + 0];
const float img_width = d_img_info_vec[5 * image_index + 1];
const float min_size_scaled = min_size * d_img_info_vec[5 * image_index + 2];
```
...at which point one would expect `5 * image_index + 3` to be assigned to some `foo` and a `5 * image_index + 4` to some `bar`, but this never happens. These two float elements effectively ending up being nothing more than some weird unused padding, it seems.

I suspect they were used as extra attributes at some earlier development stage, and code wasn't properly updated upon their removal?

@aaroey @pengchongjin 
Any reason why this didn't seem like cause for concern during (pre-merge) review?

wbudd replied:
Actually, I guess the padding might be intended as a bank conflict avoidance tactic. Either way, implementation details shouldn't distort the public API; and if they do anyway there should be some comment/documentation mentioning the underlying reasoning.
